Release/1.2.0

oguzhanunlu opened the pull request.

oguzhanunlu commented:
Hi @benjben , thanks, I added release-related commits just now. There is a problem with tests, as you can see from the travis build. I investigate them now and rc1 will be out with some delay.

benjben commented on CHANGELOG:
```diff
@@ -1,3 +1,7 @@
+1.2.0 (2020-04-24)
+-----------------------------------
```
Missing a few `----`

oguzhanunlu replied:
thanks, fixed now

chuwy replied:
I don't think we were missing. Usually we aligh dashes with release name, like:

```
1.2.0 (2020-04-24)
------------------
```

But also we usually call it `Version 1.2.0`, which got lost after extraction, hence they're shorter: https://github.com/snowplow-incubator/iglu-server/blob/master/CHANGELOG#L1-L4

benjben replied:
Oops indeed I got confused by the wrong alignment, my mistake !

chuwy commented on CHANGELOG:
```diff
@@ -1,3 +1,7 @@
+---------------------------------------
+Bump Scala Common Enrich to 1.1.0 (#29)
+
 1.1.0 (2020-02-??)
```
Can you please create a ticket to fix this, @oguzhanunlu.

benjben replied:
We could take the opportunity to fix all the dashes of the file at the same time.

chuwy commented on src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/Enrich.scala:
```diff
@@ -116,9 +116,7 @@ object Enrich {
-        .collect { case Validated.Invalid(badRows) => badRows.toList }
-        .withName("flatten-bad-rows")
-        .flatten
+        .collect { case Validated.Invalid(br) => br }
```
```suggestion
        .collect { case Validated.Invalid(row) => row }
```

benjben commented:
This repo is using git-flow, #15 is already in `develop` in https://github.com/snowplow/beam-enrich/commit/a2353e1b0aeec039c969628dc0e8324bf6537e1d

oguzhanunlu commented:
Hi @benjben , @chuwy , I've addressed your feedback and intentionally put in separate commits. Please let me know how it looks, thanks in advance.

@chuwy , as Ben said, commit of #15 is put in develop after last release, so I added it to CHANGELOG

benjben commented on CHANGELOG:
```diff
@@ -1,12 +1,19 @@
+1.2.0 (2020-04-24)
+---------------------------------------
```
The hyphens are to be aligned with the release name, as stated [here](https://github.com/snowplow/beam-enrich/pull/31#discussion_r414777746)

benjben commented on src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/Enrich.scala:
```diff
@@ -184,7 +185,7 @@ object Enrich {
     enrichmentConfs: List[EnrichmentConf],
     cachedFiles: DistCache[List[Either[String, String]]]
-  ): SCollection[Validated[List[String], EnrichedEvent]] =
+  ): SCollection[Validated[String, EnrichedEvent]] =
```
```suggestion
  ): SCollection[Validated[BadRow, EnrichedEvent]] =
```

#21 is not taken care of

benjben commented on src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/Enrich.scala:
```diff
@@ -263,7 +264,7 @@ object Enrich {
     enrichmentRegistry: EnrichmentRegistry[Id],
     client: Client[Id, Json]
-  ): List[Validated[List[String], EnrichedEvent]] = {
+  ): List[Validated[String, EnrichedEvent]] = {
```
```suggestion
  ): List[Validated[BadRow, EnrichedEvent]] = {
```

benjben commented on src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/Enrich.scala:
```diff
@@ -104,7 +107,7 @@ object Enrich {
       sc.withName("raw-from-pubsub").pubsubSubscription[Array[Byte]](config.raw)
 
-    val enriched: SCollection[Validated[List[String], EnrichedEvent]] =
+    val enriched: SCollection[Validated[BadRow, EnrichedEvent]] =
```
Can't [1](https://github.com/snowplow/beam-enrich/blob/master/src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/utils.scala#L112) and [2](https://github.com/snowplow/beam-enrich/blob/master/src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/utils.scala#L133) return `BadRow` as well ?

oguzhanunlu replied:
Right, thanks for spotting Ben! fixing now

benjben commented on src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/utils.scala:
```diff
@@ -127,10 +126,11 @@ object utils {
    * @return a bad row (JSON string) where the line is 10 times less than the max size
```
```suggestion
   * @return a bad row where the line is 10 times less than the max size
```

benjben commented on src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/utils.scala:
```diff
@@ -109,15 +109,14 @@ object utils {
     maxSizeBytes: Int,
     processor: Processor
-  ): String = {
+  ): BadRow = {
```
Scala doc needs to be updated (`a bad row (JSON string)` => `a bad row`)

chuwy commented:
Could you please first squash all aux commits, @oguzhanunlu.

chuwy commented on src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/Enrich.scala:
```diff
@@ -43,6 +43,9 @@ import utils._
 
   private val logger = LoggerFactory.getLogger(this.getClass)
+
+  implicit val badRowScioCodec: Coder[BadRow] = Coder.kryo[BadRow]
```
Are we sure it will be enough? @dilyand do you remember what our workaround included?

benjben replied:
If it can compile it should be. When I tried to do it few months ago compiling was always resulting in OOM.

chuwy replied:
IIRC we then hit runtime exception because Beam couldn't serialize `Map`. @dilyand probably remembers how we fixed this.

Edit: runtime exception was in tests.

dilyand replied:
If I remember correctly, the serialization issue only comes to light if we use `BadRow` in a `groupBy`. 

But anyway, the solution was a custom mapper: https://github.com/snowplow-product/gcp-bad-row-aggregator/blob/master/src/main/scala/com/snowplowanalytics/monitoring/googlecloud/badrows/aggregates/DMapper.scala

oguzhanunlu replied:
thanks Dilyan!

benjben commented:
LGTM @oguzhanunlu !
